### PR TITLE
Additional patches to eLua

### PR DESCRIPTION
**Pull request by ThomasHornschuh**
These is a collection of all patches I made over the last months.

There are around three topics:
a) UART and XMODEM
When I implemented eLua on my FPGA platform I run into overrun problems when using the recv command with high data rates (500.000bit/s). I was successfully able to tune the coding a bit: I changed to code to only check timers when there is no char available (so always call platform_s_uart_recv with a timeout of 0 first).
In addition I added support for xmodem 1K. This is not so much relevant for eLua, where downloaded scripts are usually small, but it adds convenience. Maybe my default buffer setting of 4K  maybe to large for platform tight on memory. It would be better to make it configurable in the board definition file.

b) virtual timer
I found a few minor bugs in the virtual timer implementation. 

c)Bit library
The library didn't work with numbers >=0x80000000, at least when eLua is compiled with doubles as numbers. This reason is that the  C library helper function which converts a double to a signed 32 bit integer is setting the result to +MAXINT when the value of the double is larger then +MAXINT.
So 0x80000000 is converted to 0x7FFFFFFF. The statemment
`print(string.format("%x",bit.band(0x80000000,0x80000000))) `
resulted output "7fffffff" instead of 80000000.
As a first workaround I just changed the bit library to use unsigned ints. This looks more "natural" at a first glance, but as also drawbacks, e.g. that lua compiled with integers as numbers will behaves differently (although not incorrect).

I have also tried to use [http://bitop.luajit.org/](http://bitop.luajit.org/). It worked well, but it was not compatible with the cpu.r/w functions, because they expect unsigned numbers and failed when converting a negative  double number to an unsigned 32 Bit int , so I switched back to my first patch.

I checked the bit32 module in Lua 5.2. It works returns unsigned  numbers, but it accepts positive and negative numbers and also numbers larger then 2^32. It uses only the lower 32Bit of such numbers which basically, which is effectively treating applying a modulo 32 operation to every inbound numbers.
This is the most intuitive looking approach in my opinion, although it also lacks consistency between floating point and integer builds. 

At the end I'm not so happy with the current state of the bit library, because it doesn't deal with well with input values outside the range 0..2^32-1. What do you think?

### CODE REVIEW

File: src/common_tmr.c [hunk 1]
```diff
@@ -79,6 +79,10 @@ static void vtmr_reset_timer( unsigned vid )
   unsigned id = VTMR_GET_ID( vid );
 
   vtmr_reset_idx = ( s8 )id;
+  
+  // TH: Ensure that Interrupts are enabled before timer is reset, otherwise eLua will hang forever....
+  platform_cpu_set_global_interrupts(PLATFORM_CPU_ENABLE);
+  // End TH
```
Comment (bogdanm) [inline src/common_tmr.c]: The usual pattern for interrupt manipulation (which restores the previous state of interrupts) is:

```
int oldstate = platform_cpu_set_global_interrupts(PLATFORM_CPU_ENABLE);
// do your stuff here
platform_cpu_set_global_interrupts(oldstate);
```

**Comment by bogdanm** [comment]
Thanks for the PR. I understand the issues with the bit module, and I think your solution is likely "the most right" we can get with the current state of things. The proper solution is probably the built-in integer support in Lua 5.3 :)

**Comment by ThomasHornschuh** [comment]
Hi, I have updated the PR:
**bit module:**
 I was still not satisfied with the way it worked. Now I changed the TOBIT macro slightly that it the bit module works exactly the same way as the bit32 module in lua 5.2. Basically it works now also correct when a negative number is passed (it is treaed to the 32Bit 2th complement representation of the number, e.g.  passing -1 is treated the same as passing 0xffffffff or 2^32-1.
Numbers outside the 32Bit range are scaled down modulo 2^32.
The implementation is very simple: The double value is converted to a signed 64Bit integer and then only the lower 32Bit are used. 
I have added the same approach to the read/write functions of the cpu module.

**virtal timers:**
I have changed to coding to your recommendation. Nevertheless I'm not sure what "philospophy" you want to follow with eLua:

- The user should know what he is doing when switching of interrupts, e.g. with cpu.cli() which means that it will crash the system when using functions which depend on enabled interrupts
-  eLua is protecting the user, at least to the extent possible fromdoing  silly things. 

Basically disabling interrupts globally will also hang the net module or the uart/console when buffering is used...


BTW: Is there any active mailing list for eLua for discussing? I'm still considering porting lwIP to eLua, which may lead to questions, especially regarding the build scripts.

**Comment by bogdanm** [comment]
> Now I changed the TOBIT macro slightly that it the bit module works exactly the same way as the bit32 module in lua 5.2.

Personally, I'm very satisfied with this.

> Nevertheless I'm not sure what "philospophy" you want to follow with eLua:

The whole "enable/disable and restore previous state" thing is a very simplistic way to resolve synchornization-specific issues that are exacerbated in eLua for two reasons: the common code is platform-independent and there is no RTOS. The mechanism can definitely be improved, but still there are limits to what one can do with interrupts without having to worry howthe interrupt controller works on the underlying hardware. The current idea is to disable interrupts in critical sections for as little time as possible, then enable them back (or rather restore their previous state). The mechanism has obvious limitations, but at this time I don't have plans to change it.

> BTW: Is there any active mailing list for eLua for discussing?

http://elua-development.2368040.n2.nabble.com/ is being actively monitored by a number of people, including myself.
